test/e2e/volume_provisioning.go

```diff
@@ -125,6 +126,56 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
 
 			testDynamicProvisioning(c, claim)
 		})
+
+		It("should test that deleting a claim before the volume is provisioned deletes the volume. [Volume]", func() {
+			// This case tests for the regressions of a bug fixed by PR #21268
+			// REGRESSION: Deleting the PVC before the PV is provisioned can result in PV
+			// not being deleted
+
+			// Given the small race window, attempt to produce it iteratively
+			var raceAttempts int = 100
+
+			By("creating a StorageClass")
+			class := newStorageClass()
+			class.Name = ns + "-sc" // pseudo-random Storage Class name
+			class, err := c.Storage().StorageClasses().Create(class)
+			defer c.Storage().StorageClasses().Delete(class.Name, nil)
+			Expect(err).NotTo(HaveOccurred())
+
+			By(fmt.Sprintf("Creating and deleting PersistentVolumeClaims for %d iterations", raceAttempts))
+			claim := newClaim(ns, false)
+			claim.Annotations[storageutil.StorageClassAnnotation] = class.Name
+			for i := 0; i < raceAttempts; i++ {
+				tmpClaim, err := c.Core().PersistentVolumeClaims(ns).Create(claim)
+				Expect(err).NotTo(HaveOccurred())
+				err = c.Core().PersistentVolumeClaims(ns).Delete(tmpClaim.Name, nil)
+				Expect(err).NotTo(HaveOccurred())
+			}
+
+			By(fmt.Sprintf("Checking for residual PVs associated with StorageClass %s", class.Name))
+			clusterPVs, err := c.Core().PersistentVolumes().List(metav1.ListOptions{})
+			regressedPVs := make([]v1.PersistentVolume, 0)
+			for _, pv := range clusterPVs.Items {
```

Comment:
Deleting a PV might take some time, it's not guaranteed to be done when `c.Core().PersistentVolumeClaims(ns).Delete(tmpClaim.Name)` returns. You need add another loop, trying several times with a sleep between the attempts, trying for several minutes:

```go
for i := 0; i < 30 && len(regressedPVs) > 0; i++ {
  By(fmt.Sprintf("Checking for residual PVs associated with StorageClass %s", class.Name))
  ...
  time.Sleep(10)
}